### DISCUSSION TITLE
Should Directus require old password when setting new password

### DISCUSSION THREAD

**Discussion by apstanisic** [directus/directus#3508] [Archive]
Currently only requirement is to provide new password.

**Comment by benhaynes** [comment]
An interesting idea — though this would require some additional logic/conditionals for admins to reset passwords.

**Comment by apstanisic** [comment]
It should probably be that admins can change password for normal users without needing old password, but normal users and admins can't change their own password without old password. And should admin be able to change password for other admin user?

**Comment by rijkvanzanten** [comment]
To be fair, you have to be logged in in order to change your password, which means you've effectively provided your old password at one point during the update flow 🤔 (unless you're going through the email-reset flow, but that's not what we're discussing here).

An easier solution would be to prevent changing passwords through the app altogether, and solely rely on the reset-password flow. This would restrict admins from setting other users' passwords, but that might also be considered a security perk

**Reply by apstanisic** [reply]
I think this will explain reasons better than I: [https://security.stackexchange.com/questions/162021/what-is-the-purpose-of-confirming-old-password-to-create-a-new-password](https://security.stackexchange.com/questions/162021/what-is-the-purpose-of-confirming-old-password-to-create-a-new-password). 
Password rest flow is always going to be problem if email is compromised, but there is nothing we can do about that.

Also, with having password change in same form as user info, user can accidentally type some string and change password without realizing. Maybe some auto-fill extensions fills this field without user even realising.

Yeah, reset-password flow is a solid solution. 

One option for admins ability to set user passwords, is that we have expiring passwords. We can have `expires_at` field that say when will password expire, and `null` will be never. That way, admin can set that password expires in the past, and next time user logs in, he will be prompted to change password. That way, admin can't set password manually, for user to change password, and we have additional perk of password expiration, that a lot of enterprise firms use. Something similar exist in Linux.

Also of topic, but it would be cool, if we offered ability to user to logout other sessions after changing password, or even without changing passwords.  Because currently when user password is stolen, he can't invalidate other sessions. Here is how NextCloud does it.
![password-sessions](https://user-images.githubusercontent.com/19491605/103377185-efb8e880-4ade-11eb-9a47-d2a7ca83657f.png)
